### PR TITLE
Clean up legacy

### LINKED ISSUE #633: remove legacy bundle service

### PR DESCRIPTION
**Pull request by kashizui**
- [x] (Merge with existing legacy cleanup branch w/e that is doing)
- [x] Move everything out of BundleService into proper modules in the rest package (web terminal)
- [x] Keep existing legacy APIs used by the frontend in one neat package that can be slowly deprecated
- [x] Grep for all the legacy kwargs and TODOs and FIXMEs
- [x] replace RPC calls in Cli
- [x] Simplify the multiple (cyclical) levels of indirection in worksheet interpretation
- [x] Remove base64 encoded fields that are immediately decoded.
- [x] check that cat_target file stuff is necessary
- [x] better error handling
- [x] regenerate REST docs
- [x] remove already migrated APIs
- [x] remove deprecated routes
- [x] address comments
- [x] human-readable REST module names in docs
- [x] fill out some more rest docs

Fixes #633

### CODE REVIEW

**Comment by kashizui** [comment]
@percyliang basically complete and ready for review

File: codalab/rest/cli.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+from codalab.lib import (
+    bundle_cli,
+    worksheet_util,
+)
+from codalab.lib.codalab_manager import CodaLabManager
+from codalab.objects.oauth2 import OAuth2Token
+
+
+@post('/cli/command')
+@post('/api/worksheets/command/')  # DEPRECATED ROUTE
```
Comment (percyliang) [inline codalab/rest/cli.py]: Do we still need to keep this route?

File: codalab/rest/cli.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+def post_worksheets_command():
+    # TODO(klopyrev): The Content-Type header is not set correctly in
+    # editable_field.jsx, so we can't use request.json.
+    data = json.loads(request.body.read())
+
+    if data.get('raw_command', None):
+        data['command'] = worksheet_util.get_worksheet_info_edit_command(data['raw_command'])
+
+    if not data.get('worksheet_uuid', None) or not data.get('command', None):
```
Comment (percyliang) [inline codalab/rest/cli.py]: In all of this, don't need the `None` argument to get since that's the default.

File: codalab/rest/cli.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+
+@post('/cli/command')
+@post('/api/worksheets/command/')  # DEPRECATED ROUTE
+def post_worksheets_command():
+    # TODO(klopyrev): The Content-Type header is not set correctly in
+    # editable_field.jsx, so we can't use request.json.
+    data = json.loads(request.body.read())
+
+    if data.get('raw_command', None):
```
Comment (percyliang) [inline codalab/rest/cli.py]: `raw_command in data`?

File: codalab/rest/cli.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        data['command'] = worksheet_util.get_worksheet_info_edit_command(data['raw_command'])
```
Comment (percyliang) [inline codalab/rest/cli.py]: This logic is kind of strange: why is raw_command only for edits?  The name wouldn't suggest that...

File: docs/rest.md [hunk 5]
```diff
@@ -10,15 +10,18 @@ subject to change at any time. Feedback through our GitHub issues is appreciated
   - [Bundle Permissions API](#bundle-permissions-api)
   - [Oauth2 API](#oauth2-api)
   - [Users API](#users-api)
-  - [Worksheet Items API](#worksheet-items-api)
+  - [Chatbox API](#chatbox-api)
+  - [Help API](#help-api)
   - [Workers API](#workers-api)
   - [Bundles API](#bundles-api)
+  - [Interpret API](#interpret-api)
```
Comment (percyliang) [inline docs/rest.md]: Can we call this `worksheet-interpretation-api` or something (to say it's for worksheets)?

File: codalab/lib/worksheet_util.py [hunk 6]
```diff
@@ -379,100 +375,6 @@ def friendly_render_dep(dep):
     return None
 
 
-def interpret_file_genpaths(client, requests):
```
Comment (percyliang) [inline codalab/lib/worksheet_util.py:382]: Do we have a way to think about which stuff goes in `worksheet_util.py` and what goes in `interpet,py`?  Maybe would be good to document this.

File: codalab/rest/cli.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+    The CLI uses JsonApiClient to communicate back to the REST API.
+    This is admittedly not ideal since now the REST API is essentially
+    making HTTP requests back to itself. Future potential solutions might
+    include creating a subclass of JsonApiClient that can reroute HTTP
+    requests directly to the appropriate Bottle view functions.
+    """
+    output_buffer = StringIO()
+    rest_client = JsonApiClient(rest_url(), get_user_token)
+    manager = CodaLabManager(
```
Comment (percyliang) [inline codalab/rest/cli.py:108]: This always makes me really nervous.  Can we double check that we didn't do anything that opens up new security holes?

**Review by percyliang — COMMENTED** [review]
Nice!

File: codalab/lib/worksheet_util.py (hunk 6 above)
Comment (kashizui) [reply codalab/lib/worksheet_util.py:382]: Good idea, I'll document it. The distinction is that worksheet_util doesn't make any calls to the model, they're kind of just like static helper functions.

File: codalab/rest/cli.py (hunk 1 above)
Comment (kashizui) [reply codalab/rest/cli.py]: This needs to be merged first, then I'll add a PR on worksheets to use the new route, then I can create a new PR to delete all the `DEPRECATED ROUTE`s. I'll add a comment somewhere or create an issue.

File: codalab/rest/cli.py (hunk 4 above)
Comment (kashizui) [reply codalab/rest/cli.py]: I actually don't know where this is used, I'll look into it and make renames/deletions when they make sense.

File: codalab/rest/cli.py (hunk 7 above)
Comment (kashizui) [reply codalab/rest/cli.py:108]: Sure thing I'll take a look.
@fabeschan would be great to get a second set of eyes on this as well :)

File: docs/rest.md (hunk 5 above)
Comment (kashizui) [reply docs/rest.md]: Just in the docs or in the API routes as well?

Comment (percyliang) [reply docs/rest.md]: Definitely in the docs.  API routes - I could go either way.  Probably fine to leave to keep it short.  (Alternatively, we could do winterpret.)

File: codalab/rest/bundles.py [hunk 8]
```diff
@@ -343,22 +344,36 @@ def _fetch_bundle_contents_info(uuid, path=''):
           "link": "<string representing target if file is a symbolic link>",
           "type": "<file|directory|link>",
           "size": <size of file in bytes>,
```
Comment (fabeschan) [inline codalab/rest/bundles.py:355]: I wonder if you also want to update this part to reflect the new parameter

File: codalab/rest/cli.py (hunk 7 above)
Comment (fabeschan) [reply codalab/rest/cli.py:108]: okay

**Comment by kashizui** [comment]
Since I removed all uses of legacy or deprecated routes from codalab/codalab-worksheets#309, I was able to completely remove the legacy module now.

Also made some modifications to the API necessary for the frontend migration and improved bundle metadata validation.

Ready for final reviews @percyliang @fabeschan @maxwang7

File: codalab/client/json_api_client.py [hunk 9]
```diff
@@ -603,17 +603,49 @@ def upload_contents_blob(self, bundle_id, fileobj=None, params=None,
...
                 progress_callback=progress_callback)
 
-    @wrap_exception('Unable to make RPC call \'{1}\'')
-    def rpc(self, method, *args, **kwargs):
-        request = {
-            'method': method,
-            'args': args,
-            'kwargs': kwargs,
-        }
+    @wrap_exception('Unable to interpret file genpaths')
```
Comment (fabeschan) [inline codalab/client/json_api_client.py:606]: Perhaps dumb question: what's a genpath?

Comment (fabeschan) [reply codalab/client/json_api_client.py:606]: (got a thorough explanation in person)

**Review by fabeschan — APPROVED** [review]
LGTM great work

File: codalab/rest/cli.py (hunk 7 above)
Comment (fabeschan) [reply codalab/rest/cli.py:108]: I think the way this is done makes sense? Unless I missed something

File: codalab/lib/worksheet_util.py (hunk 6 above)
Comment (kashizui) [reply codalab/lib/worksheet_util.py:382]: done

File: codalab/rest/bundles.py (hunk 8 above)
Comment (kashizui) [reply codalab/rest/bundles.py:355]: parameter got removed

**Pull request merged**